#3634 Create/Edit Machinery Modals

EddieTheEgg opened the pull request:
## Changes

Added create and edit machinery modal with the table of machineries in admin tools -> schedule tab.
Also when editing a machinery modal, the data inputted previously will fill as placeholders as requested from the ticket.

## Notes

Delete machinery button is disabled for now since I didn't have hooks or connect to backend endpoints for this feature.


## Screenshots
<img width="2702" height="1490" alt="Screenshot 2025-10-25 at 8 08 28 PM" src="https://github.com/user-attachments/assets/f49b0017-ce57-4b27-ba27-df0f11ffeb98" />
<img width="2024" height="1338" alt="Screenshot 2025-10-25 at 8 08 55 PM" src="https://github.com/user-attachments/assets/fbc4b230-75bd-498f-9585-dede10b16fcd" />
<img width="1386" height="1266" alt="Screenshot 2025-10-25 at 8 09 14 PM" src="https://github.com/user-attachments/assets/9cbf3301-7f70-4576-894c-09f7938a421c" />

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #3400

EddieTheEgg committed:
#3634 Added machinery create/edit modal with table in admin schedule

EddieTheEgg committed:
#3634 minor naming consistency with other machinery modals

wavehassman commented on src/frontend/src/apis/calendar.api.ts:
```diff
@@ -13,3 +13,27 @@ export const postCreateShop = (payload: { name: string; description: string }) =
+  });
+};
+
+export const postCreateMachinery = (payload: { name: string; shopId: string; quantity: number; description?: string }) => {
```
I'm confused on description being here since the modal is only name, shop and quantity. It's also in the backend

EddieTheEgg replied:
When I was creating the endpoints like createMachinery and editMachinery, I might've I mixed up interpretation from ShopMachinery and Machinery fields in the backend lol. The schema has ShopMachinery hold an optional description field but Machinery itself doesn't. I removed in the frontend for now, but if I need to update backend to remove such optional field I can do that as well (unless it has actual purpose?).

wavehassman replied:
Yeah it won’t serve any purpose for machinery since it doesn’t take in a description so you can delete it from the backend too. Just take description out of the route, controller and service for create and edit machinery.

wavehassman commented on src/frontend/src/apis/calendar.api.ts:
```diff
@@ -13,3 +13,27 @@ export const postCreateShop = (payload: { name: string; description: string }) =
+  name: string;
+  shopId: string;
+  quantity: number;
+  description?: string;
```
same here

wavehassman commented on src/frontend/src/pages/AdminToolsPage/ScheduleConfig/Machinery/CreateMachineryModal.tsx:
```diff
@@ -0,0 +1,134 @@
+interface CreateMachineryModalProps {
+  open: boolean;
+  onClose: () => void;
+  onSubmit: (data: CreateMachineFormValues) => Promise<unknown> | unknown;
```
you shouldn't be using unknown

wavehassman commented on src/frontend/src/pages/AdminToolsPage/ScheduleConfig/Machinery/CreateMachineryModal.tsx:
```diff
@@ -0,0 +1,134 @@
+import { Box, FormControl, FormHelperText, Typography, MenuItem, Select, SelectChangeEvent } from '@mui/material';
```
the create and edit modals are very similar, look at how other create/edit modals are abstracted to get rid of duplicate code

EddieTheEgg replied:
Abstracted referencing MilestoneForm structure.

wavehassman commented on src/frontend/src/pages/AdminToolsPage/ScheduleConfig/Machinery/EditMachineryModal.tsx:
```diff
@@ -0,0 +1,139 @@
+  open: boolean;
+  onClose: () => void;
+  onSubmit: (data: EditMachineFormValues) => Promise<unknown> | unknown;
+  initialValues?: EditMachineFormValues;
```
the name doesn't populate on edit

JoshuaGoldberg replied:
I think it may be a mismatched field on line 107?

wavehassman commented on src/frontend/src/pages/AdminToolsPage/ScheduleConfig/Machinery/EditMachineryModal.tsx:
```diff
@@ -0,0 +1,139 @@
+import { Box, FormControl, FormHelperText, Typography, MenuItem, Select, SelectChangeEvent } from '@mui/material';
```
editing a machine only works for the quantity. I can't get it to change the name or shop, the table does not update

wavehassman reviewed (changes requested):
Really good! Just a couple bugs and fixes

EddieTheEgg committed:
#3634 removed description fields and update frontend data populating

EddieTheEgg committed:
#3634 Updated tests without description field, and adjusted editMachinery payload sending to backend via seperated params

EddieTheEgg committed:
#3634 Updated seed to remove description inputs

EddieTheEgg committed:
#3634 Abstracted edit and create machinery modal to follow other modal abstractions similarly

wavehassman commented on src/backend/src/services/calendar.services.ts:
```diff
@@ -452,10 +442,10 @@ export default class CalendarService {
         shops: {
           updateMany: {
-            where: { shopId },
+            where: {},
```
shouldn't the shop id still be in here so we're changing the right shop machinery

EddieTheEgg replied:
Using where: {} updates all ShopMachinery records for this machinery to the new shopId and quantity. Since we scope by machineryId, each record’s shopId changes to the new value. Using where: { shopId } in this scenario is like finding any machinery that already has that new shop we want to update to, which I don't think make sense in this case. Correct me if I'm wrong though. I did test with keeping shopId in where:{} versus not and what I have currently updates the table correctly while the where: {shopId} does not update shop.

wavehassman replied:
As long as we are getting the expected functionality, that is ok. I would just double check or add some tests to make sure it is updating correctly.

wavehassman reviewed (changes requested):
This looks great! Just one clarification

EddieTheEgg committed:
#3634 Added one more machinery test to make sure edit works

chpy04 commented on src/frontend/src/pages/AdminToolsPage/ScheduleConfig/AdminToolsScheduleConfig.tsx:
```diff
@@ -105,14 +111,76 @@ const AdminToolsScheduleConfig: React.FC = () => {
+                    let machineQuantity = '—';
+                    if (machine.shops && machine.shops.length > 0) {
+                      shopName = machine.shops[0].shop.name;
+                      machineQuantity = machine.shops[0].quantity.toString();
```
if a machine is connected to multiple shops we can't just only show the first one. there should be a row for each shop for each machine

chpy04 commented on src/frontend/src/pages/AdminToolsPage/ScheduleConfig/Machinery/EditMachineryModal.tsx:
```diff
@@ -0,0 +1,29 @@
+  const { isLoading, isError, error, mutateAsync } = useEditMachinery(machinery.machineryId);
+
+  const machineryData: MachineryFormValues = {
+    shopId: machinery.shops?.[0]?.shop?.shopId || '',
```
same thing here, the shop should also come from the form and these values will be for that specific relationship

EddieTheEgg committed:
#3634 Updated machinery backend/frontend relationship one-to-many when adding machine and assigning to multiple shops

EddieTheEgg committed:
#3634 minor quantity change to accept text input instead of hardcoded list of nums

EddieTheEgg commented:
Had to add two new params to editMachinery with the new machinery relationship (machine can be connected to multiple shops) so we can identify which shop-machinery relationship is being edited (so Waverly's comment earlier about where: { shopId} IS actually necessary for these shop-machinery relationships. Adjusted the service for editMachinery endpoint so that it handles duplicate machinery names with different shop assignment, or merging two machines into one shop and adjusting the quantity etc. Adjusted tests to adapt to these service changes and also tested on the website it merges/makes the separate rows as you said nicely!

Also minor frontend change with quantity input, cause I realize if we keep merging machines together there's a likely chance machines go over 10 which will visually error when trying to edit that specific machine quantity. I made the quantity input TextField, more flexible compared to just hardcoding inputs up to 10.

chpy04 commented on src/backend/src/services/calendar.services.ts:
```diff
@@ -412,14 +459,13 @@ export default class CalendarService {
     organization: Organization,
-    description?: string
+    originalShopId?: string,
+    shopMachineryId?: string
```
we shouldn't be taking in a shopmachineryId here because these are always defined by their associated shop and machinery

chpy04 reviewed (changes requested):
Good work thinking through all the cases here and I believe this looks right, although in terms of readability I want to seperate the cases into different endpoints so it is very clear what is an edit and what is a creation. For the create and edit machinery endpoints, I think these should only work with the name, and should be completely unrelated to the shops and quantities in those shops. From there, there should be a seperate endpoint like "add-machinery-to-shop", which takes in a quantity as well as a machinery and shop id, and then sets that quantity to the shopmachinery if one exists or creates a new one otherwise. This could also serve as the delete by checking if the quantity is 0 and deleting the shopmachinery if it is

EddieTheEgg committed:
#3634 restructured endpoints to make machine first before doing something with it to shops

EddieTheEgg committed:
Merge branch 'feature/calendar-improvements' into #3634-create-edit-machinery-modal-edison-kwok

EddieTheEgg committed:
#3634 Attempt to fix merge conflict and recommit new endpoint changes

EddieTheEgg committed:
Merge branch '#3634-create-edit-machinery-modal-edison-kwok' of https://github.com/Northeastern-Electric-Racing/FinishLine into #3634-create-edit-machinery-modal-edison-kwok

EddieTheEgg committed:
#3634 fixing build issues

EddieTheEgg commented:
I refactored machinery endpoints to separate concerns: createMachinery and editMachinery now only handle machinery names (returns/reuses existing machinery by name). Added a new addMachineryToShop endpoint that manages shop-machinery relationships separately and does the actual edge cases we discussed earlier. The only thing I changed based on what you said was with quantity being 0. For adding machine, the machine quantity must be >=1, but for editing user can input 0 and it will delete the existing machinery/shop when trying to "addMachineryToShop". If you need to update this quantity logic let me know. Overall the service should be delegated much better with addMachineryToShop handling any connection between shops and machinery, while machinery have their own endpoints.

Also ignore the merge commit, accidentally ran wrong git merge command. I undid so that my branch and features/calendar-improvement are back to their original version (Unless I did this wrong oops).

EddieTheEgg committed:
#3634 Updated machinery edit to only allow delete machinery via delete and not have quantity 0, and for relationships not actual machines

EddieTheEgg committed:
#3634 Fixed linting issues and added relationship adjustment to quantity merging ShopMachineries

EddieTheEgg commented:
New fixes, added the quantity 0 feature to fully adjust for merging as well (Merging as in deleting any relationship with the same name and assigned shop to the edited ShopMachinery).
